File: packages/mobile-core/lib/src/utils/format.dart
import '../models/models.dart';

/// Pemformatan angka & waktu.
///
/// Semua angka ditampilkan dengan JetBrains Mono — itulah yang memberi kesan
/// "readout konsol". Fungsi di sini menjaga bentuknya konsisten di kedua app.

/// 1240 -> "1.2 KM" · 850 -> "850 M"
String formatDistance(int? meters) {
  if (meters == null) return '—';
  if (meters < 1000) return '$meters M';
  return '${(meters / 1000).toStringAsFixed(1)} KM';
}

/// 420 -> "7 MNT"
String formatDuration(int? seconds) {
  if (seconds == null) return '—';
  final minutes = (seconds / 60).round().clamp(1, 1 << 30);
  if (minutes < 60) return '$minutes MNT';
  final h = minutes ~/ 60;
  final m = minutes % 60;
  return m == 0 ? '$h JAM' : '$h JAM $m MNT';
}

/// Versi panjang untuk readout hero: 420 -> "7 MENIT"
String formatDurationLong(int? seconds) {
  if (seconds == null) return '—';
  final minutes = (seconds / 60).round().clamp(1, 1 << 30);
  return '$minutes MENIT';
}

/// "BARU SAJA" · "2 MNT LALU" · "3 JAM LALU"
String formatRelative(DateTime time) {
  final diff = DateTime.now().difference(time);
  if (diff.inMinutes < 1) return 'BARU SAJA';
  if (diff.inMinutes < 60) return '${diff.inMinutes} MNT LALU';
  if (diff.inHours < 24) return '${diff.inHours} JAM LALU';
  return '${diff.inDays} HARI LALU';
}

/// "12.08.2026 · 14:20" — format tanggal di layar Riwayat.
String formatDateTime(DateTime t) {
  String p(int n) => n.toString().padLeft(2, '0');
  return '${p(t.day)}.${p(t.month)}.${t.year} · ${p(t.hour)}:${p(t.minute)}';
}

/// Inisial avatar: "Ahmad Ridwan" -> "AR"
String initials(String name) {
  final parts = name.trim().split(RegExp(r'\s+')).where((s) => s.isNotEmpty).toList();
  if (parts.isEmpty) return '?';
  if (parts.length == 1) {
    return parts.first.substring(0, parts.first.length.clamp(0, 2)).toUpperCase();
  }
  return (parts.first[0] + parts.last[0]).toUpperCase();
}

/// Koordinat gaya konsol: "-6.5971, 106.8060"
String formatCoords(LatLngPoint p) =>
    '${p.lat.toStringAsFixed(4)}, ${p.lng.toStringAsFixed(4)}';

/// Selisih dua waktu dalam menit — dipakai statistik waktu respons.
int? minutesBetween(DateTime? a, DateTime? b) {
  if (a == null || b == null) return null;
  return b.difference(a).inMinutes;
}

File: apps/web-dashboard/src/lib/format.ts
/**
 * Pemformatan angka & waktu.
 *
 * Semua angka di UI ditampilkan dengan JetBrains Mono — itulah yang memberi
 * kesan "readout konsol" di sistem desain ini. Fungsi di sini memastikan
 * bentuk angkanya konsisten di seluruh halaman.
 */

/** 1240 -> "1.2 KM", 850 -> "850 M" */
export function formatDistance(meters: number | null | undefined): string {
  if (meters === null || meters === undefined) return '—';
  if (meters < 1000) return `${Math.round(meters)} M`;
  return `${(meters / 1000).toFixed(1)} KM`;
}

/** 420 -> "7 MNT" */
export function formatDuration(seconds: number | null | undefined): string {
  if (seconds === null || seconds === undefined) return '—';
  const minutes = Math.max(1, Math.round(seconds / 60));
  if (minutes < 60) return `${minutes} MNT`;
  const h = Math.floor(minutes / 60);
  const m = minutes % 60;
  return m === 0 ? `${h} JAM` : `${h} JAM ${m} MNT`;
}

/** "BARU SAJA", "2 MNT LALU", "3 JAM LALU" — sesuai kartu SOS di mockup. */
export function formatRelative(iso: string): string {
  const diffMs = Date.now() - new Date(iso).getTime();
  const minutes = Math.floor(diffMs / 60_000);
  if (minutes < 1) return 'BARU SAJA';
  if (minutes < 60) return `${minutes} MNT LALU`;
  const hours = Math.floor(minutes / 60);
  if (hours < 24) return `${hours} JAM LALU`;
  const days = Math.floor(hours / 24);
  return `${days} HARI LALU`;
}

/** "12.08.2026 · 14:20" — format tanggal di layar Riwayat. */
export function formatDateTime(iso: string): string {
  const d = new Date(iso);
  const pad = (n: number) => String(n).padStart(2, '0');
  return `${pad(d.getDate())}.${pad(d.getMonth() + 1)}.${d.getFullYear()} · ${pad(d.getHours())}:${pad(d.getMinutes())}`;
}

/** Inisial untuk avatar: "Ahmad Ridwan" -> "AR" */
export function initials(name: string): string {
  const parts = name.trim().split(/\s+/).filter(Boolean);
  if (parts.length === 0) return '?';
  if (parts.length === 1) return parts[0]!.slice(0, 2).toUpperCase();
  return (parts[0]![0]! + parts[parts.length - 1]![0]!).toUpperCase();
}

/** Koordinat gaya konsol: "-6.5971, 106.8060" */
export const formatCoords = (lat: number, lng: number) =>
  `${lat.toFixed(4)}, ${lng.toFixed(4)}`;

/** Selisih dua timestamp dalam menit — dipakai statistik waktu respons. */
export function minutesBetween(a: string | null, b: string | null): number | null {
  if (!a || !b) return null;
  return Math.round((new Date(b).getTime() - new Date(a).getTime()) / 60_000);
}
